# Cleaning the dataset from Epicurious.com for SmartRecipes (a recipe recommender)

# Table of Contents
- [Cleaning the dataset from Epicurious.com for the Recipe Recommendation System](#cleaning-the-dataset-from-epicurious.com-for-the-recipe-recommendation-system)
  - [Introduction](#introduction)
  - [Load the data](#load-the-data)
  - [Data Dictionary](#data-dictionary)
  - [Drop columns](#drop-columns)
  - [Convert data types](#convert-data-types)
  - [Check for null values](#check-for-null-values)
  - [Check for duplicates](#check-for-duplicates)
  - [Save the cleaned dataset](#save-the-cleaned-dataset)


## Introduction

We have decided to proceed with the dataset at https://www.kaggle.com/datasets/hugodarwood/epirecipes.

The file of interest to us is `full_format_recipes.json`. 

In this notebook, we will clean the dataset:
- remove columns that are not relevant to us
- remove null and duplicate values

Let's start by loading the dataset.

## Load the data

In [6]:
# import relevant libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [7]:
df_json = pd.read_json('../data/raw/full_format_recipes.json')
df_json.shape

(20130, 11)

The dataset has 20130 rows and 11 columns

In [8]:
df_json.head()

directions   fat  \
0  [1. Place the stock, lentils, celery, carrot, ...   7.0   
1  [Combine first 9 ingredients in heavy medium s...  23.0   
2  [In a large heavy saucepan cook diced fennel a...   7.0   
3  [Heat oil in heavy large skillet over medium-h...   NaN   
4  [Preheat oven to 350°F. Lightly grease 8x8x2-i...  32.0   

                       date  \
0 2006-09-01 04:00:00+00:00   
1 2004-08-20 04:00:00+00:00   
2 2004-08-20 04:00:00+00:00   
3 2009-03-27 04:00:00+00:00   
4 2004-08-20 04:00:00+00:00   

                                          categories  calories  \
0  [Sandwich, Bean, Fruit, Tomato, turkey, Vegeta...     426.0   
1  [Food Processor, Onion, Pork, Bake, Bastille D...     403.0   
2  [Soup/Stew, Dairy, Potato, Vegetable, Fennel, ...     165.0   
3  [Fish, Olive, Tomato, Sauté, Low Fat, Low Cal,...       NaN   
4  [Cheese, Dairy, Pasta, Vegetable, Side, Bake, ...     547.0   

                                                desc  protein  rating  \
0                                               None     30.0   2.500   
1  This uses the same ingredients found in boudin...     18.0   4.375   
2                                               None      6.0   3.750   
3  The Sicilian-style tomato sauce has tons of Me...      NaN   5.000   
4                                               None     20.0   3.125   

                                         title  \
0              Lentil, Apple, and Turkey Wrap    
1  Boudin Blanc Terrine with Red Onion Confit    
2                Potato and Fennel Soup Hodge    
3             Mahi-Mahi in Tomato Olive Sauce    
4                    Spinach Noodle Casserole    

                                         ingredients  sodium  
0  [4 cups low-sodium vegetable or chicken stock,...   559.0  
1  [1 1/2 cups whipping cream, 2 medium onions, c...  1439.0  
2  [1 fennel bulb (sometimes called anise), stalk...   165.0  
3  [2 tablespoons extra-virgin olive oil, 1 cup c...     NaN  
4  [1 12-ounce package frozen spinach soufflé, th...   452.0

## Data Dictionary

| Column      | Non-Null Count | Dtype               | Description                                     |
|-------------|----------------|---------------------|-------------------------------------------------|
| ------      | -------------- | -----               | ------                                          |
| title       | 15969 non-null | object              | Title of the recipe                             |
| directions  | 15969 non-null | object              | Steps for the recipe                            |
| ingredients | 15969 non-null | object              | Ingredients plus description of how to cut them |
| categories  | 15969 non-null | object              | Array of categories                             |
| calories    | 15969 non-null | float64             | Calories                                        |
| rating      | 15969 non-null | float64             | Rating on a scale of 0 to 5                     |
| desc        | 10636 non-null | object              | Extra tidbits about the recipe etc.             |
| date        | 15969 non-null | datetime64[ns, UTC] | Date the recipe was created                     |
| sodium      | 15967 non-null | float64             | Sodium content                                  |
| fat         | 15901 non-null | float64             | Fat content                                     |
| protein     | 15922 non-null | float64             | Protein content                                 |

## Drop columns

Columns that are relevant to us:

`title`, `directions`, `ingredients`, `categories`, `calories`, `rating`

We will drop the other columns

In [9]:
cols_to_drop = ["desc",
                "date",
                "sodium",
                "fat",
                "protein"]

df_small = df_json.drop(columns=cols_to_drop)

In [10]:
df_small.shape

(20130, 6)

After dropping columns, we have 20130 rows and 6 columns.

In [11]:
df_small.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20130 entries, 0 to 20129
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   directions   20111 non-null  object 
 1   categories   20111 non-null  object 
 2   calories     15976 non-null  float64
 3   rating       20100 non-null  float64
 4   title        20111 non-null  object 
 5   ingredients  20111 non-null  object 
dtypes: float64(2), object(4)
memory usage: 943.7+ KB


In [12]:
df_small.sample(5)

directions  \
14154  [Preheat oven to 350°F. Sprinkle fish on both ...   
15356  [Combine buttermilk, hot sauce, and 2 tsp. sal...   
19051  [Heat oil in heavy large pot over medium-high ...   
7386   [Cook bacon in a 10-inch skillet over medium h...   
17059  [In a large saucepan combine the black beans w...   

                                              categories  calories  rating  \
14154  [Fish, Tomato, Bake, Seafood, Snapper, Leek, F...     256.0   3.750   
15356  [Sandwich, Chicken, Kid-Friendly, Lunch, Cabba...    2448.0   4.375   
19051  [Soup/Stew, Onion, Tomato, turkey, Sauté, Stew...     345.0   4.375   
7386   [Salad, Tomato, Appetizer, Quick & Easy, Dinne...     381.0   3.750   
17059  [Salad, Blender, Bean, Garlic, Side, Vegetaria...     519.0   3.125   

                                                   title  \
14154               Red Snapper with Leeks and Tomatoes    
15356  Crispy Chicken Sandwich with Buttermilk Slaw a...   
19051        Turkey-Tomato Stew with Onions and Raisins    
7386   Beefsteak Tomato, Butterhead Lettuce, and Baco...   
17059    Three-Bean Salad with Coriander Chili Dressing    

                                             ingredients  
14154  [4 5- to 6-ounce red snapper fillets, 2 tables...  
15356  [1 cup buttermilk, 1 tablespoon hot sauce (pre...  
19051  [6 tablespoons olive oil, 4 cups chopped onion...  
7386   [6 bacon slices, chopped, 1/2 cup sour cream, ...  
17059  [1 pound dried black beans, picked over, 1 pou...

## Convert data types

Some of the columns have lists in them, in order to use the CountVectorizer during preprocessing, we need these to be strings. 

We will convert the contents of the following columns from list to string:
- `directions`
- `categories`
- `ingredients`

In [13]:
print(f"Directions contains entries of type: {type(df_small['directions'][0])}")
print(f"Categories contains entries of type: {type(df_small['categories'][0])}")
print(f"Title contains entries of type: {type(df_small['title'][0])}")
print(f"Ingredients contains entries of type: {type(df_small['ingredients'][0])}")

Directions contains entries of type: <class 'list'>
Categories contains entries of type: <class 'list'>
Title contains entries of type: <class 'str'>
Ingredients contains entries of type: <class 'list'>


Let's convert these list types into strings

In [14]:
df_small['directionsStr'] = df_small['directions'].astype(str)
df_small['categoriesStr'] = df_small['categories'].astype(str)
df_small['ingredientsStr'] = df_small['ingredients'].astype(str)


In [15]:
print(f"Directions contains entries of type: {type(df_small['directionsStr'][0])}")
print(f"Categories contains entries of type: {type(df_small['categoriesStr'][0])}")
print(f"Title contains entries of type: {type(df_small['title'][0])}")
print(f"Ingredients contains entries of type: {type(df_small['ingredientsStr'][0])}")

Directions contains entries of type: <class 'str'>
Categories contains entries of type: <class 'str'>
Title contains entries of type: <class 'str'>
Ingredients contains entries of type: <class 'str'>


We also noticed an extra space at the end of the contents of `title`. Let's remove that.

In [16]:
df_small['title'][0]

'Lentil, Apple, and Turkey Wrap '

In [17]:
# Remove extra space from the end of the titles
df_small['title'] = df_small['title'].str.strip()

In [18]:
# verify that the space was removed
df_small['title'][0]

'Lentil, Apple, and Turkey Wrap'

Now that we have verified that the conversions look good, we will drop the original columns.

In [19]:
# drop the original columns that contained list types
df_small.drop(columns=['directions','categories','ingredients'], inplace=True)

In [20]:
df_small.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20130 entries, 0 to 20129
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   calories        15976 non-null  float64
 1   rating          20100 non-null  float64
 2   title           20111 non-null  object 
 3   directionsStr   20130 non-null  object 
 4   categoriesStr   20130 non-null  object 
 5   ingredientsStr  20130 non-null  object 
dtypes: float64(2), object(4)
memory usage: 943.7+ KB


## Check for null values

In [21]:
df_small.isna().sum().sum()

4203

There are 4203 missing values.

In [22]:
df_small.isna().sum()

calories          4154
rating              30
title               19
directionsStr        0
categoriesStr        0
ingredientsStr       0
dtype: int64

Let's drop these rows that have null values

In [23]:
df_small = df_small.dropna(axis=0)

In [24]:
df_small.isna().sum()

calories          0
rating            0
title             0
directionsStr     0
categoriesStr     0
ingredientsStr    0
dtype: int64

In [25]:
df_small.shape

(15969, 6)

After dropping rows that have null values, we have 15969 rows and 6 columns.

## Check for duplicates

In [26]:
df_small.duplicated().sum()

1443

There are 1443 duplicates, we will drop these rows.

In [27]:
df_small.drop_duplicates(keep='first', inplace=True)

In [28]:
df_small.duplicated().sum()

0

In [29]:
df_small.shape

(14526, 6)

After dropping duplicates, we have 14526 rows and 6 columns

In [30]:
df_small.sample(5)

calories  rating                                              title  \
14837     588.0   4.375                          Blue Cheese Potato Gratin   
10185     228.0   4.375  Lemon Meringue Pie with Hazelnut Shortbread Crust   
3583      551.0   4.375     Ricotta and Spinach Tortelloni in Tomato Sauce   
6803      554.0   3.750       Crispy Noodle Cake with Shrimp and Snow Peas   
17726     208.0   5.000  Lychee Compote with Raspberries and Champagne ...   

                                           directionsStr  \
14837  ['Preheat oven to 375°F.', 'Heat oil in a 12-i...   
10185  ['Grind nuts finely in processor. Add flour an...   
3583   ['Heat 3 tablespoons oil in large skillet over...   
6803   ['In a kettle of salted boiling water cook the...   
17726  ['Lightly oil a 9-inch square baking pan and l...   

                                           categoriesStr  \
14837  ['Milk/Cream', 'Onion', 'Potato', 'Side', 'Bak...   
10185  ['Dairy', 'Egg', 'Dessert', 'Bake', 'Lemon', '...   
3583   ['Garlic', 'Leafy Green', 'Pasta', 'Tomato', '...   
6803   ['Wok', 'Ginger', 'Pasta', 'Shellfish', 'Stir-...   
17726  ['Champagne', 'Berry', 'Fruit', 'Dessert', 'Co...   

                                          ingredientsStr  
14837  ['2 tablespoons olive oil', '1 pound onions (3...  
10185  ['1/3 cup hazelnuts, toasted, husked', '1 1/4 ...  
3583   ['5 tablespoons olive oil', '1 cup chopped oni...  
6803   ['3/4 pound fresh or dried thin Asian egg nood...  
17726  ['1 cup sugar', '3 cups Champagne or other spa...

## Save the cleaned dataset

Let's save the cleaned dataset to a file called full_recipes_cleaned_2.csv

In [31]:
df_small.to_csv("../data/interim/full_recipes_cleaned_2.csv",index=True,index_label='recipeId')